### PR TITLE
Migrating to menubar application

### PR DESCRIPTION
**Pull request by danielbuechele**
This implements the discussed UI changes and makes atemOSC a menubar application.
I don't have an ATEM device at hand so I couldn't test this and would appreciate help.

### CODE REVIEW

**Comment by SteffeyDev** [comment]
Nice job, I’ll test this out on a switch tomorrow to see how it works

**Comment by sneat** [comment]
@SteffeyDev what hardware do you have access to? I (sometimes) have access to a TVS, 1ME and 2ME.

File: atemOSC/AppDelegate.mm [hunk 1]
```diff
@@ -97,6 +115,21 @@ - (void)applicationDidFinishLaunching:(NSNotification *)aNotification
...
+}
+
+- (void)showOSCAddresses
+{
+	[helpPanel makeKeyAndOrderFront:self];
+}
+
+- (void)showLogMessages
+{
+    [logPanel makeKeyAndOrderFront:self];
```
Comment (SteffeyDev) [inline atemOSC/AppDelegate.mm]: Uses spaces instead of tabs

Comment (danielbuechele) [reply atemOSC/AppDelegate.mm]: Nooooo..... ;-)

File: atemOSC/AppDelegate.h [hunk 2]
```diff
@@ -28,50 +28,47 @@
...
-	OSCManager*					manager;
-	
-	IBOutlet OSCAddressPanel*   helpPanel;
-	IBOutlet NSTextView*        logTextView;
+	OSCReceiver*                        mOscReceiver;
+	OSCManager*                         manager;
+	IBOutlet OSCAddressPanel*           helpPanel;
+	IBOutlet NSPanel*                   logPanel;
+	IBOutlet NSTextView*                logTextView;
+	NSStatusItem*                       statusItem;
```
Comment (SteffeyDev) [inline atemOSC/AppDelegate.h]: Should probably use tabs to align instead of spaces

**Comment by SteffeyDev** [comment]
When I ran for first time, I saw this warning:
```
Main Thread Checker: UI API called on a background thread: -[NSButton setImagePosition:]
PID: 88167, TID: 38628747, Thread name: (none), Queue name: com.apple.root.background-qos, QoS: 9
Backtrace:
4   atemOSC                             0x000000010000d69d -[AppDelegate switcherDisconnected] + 285
```

Should probably look into this, as these kind of errors have been known to cause crashes in the past

**Comment by SteffeyDev** [comment]
Also, when launched, if any of the preferences are missing (IP's or ports), I think the preferences panel should launch automatically so that the user can put in the needed values.  Otherwise, on first launch it might take people some time to figure out why nothing is happening (if they don't see the new icon in the status bar).

File: atemOSC/AppDelegate.mm (hunk 1 above)
Comment (sneat) [reply atemOSC/AppDelegate.mm]: Use one tab followed by two spaces ;)

**Comment by SteffeyDev** [comment]
A few more notes after testing with actual device:
* I noticed that when log messages are updated in the log window, the text flickers (disappears for a moment and comes back).  This might not be new but I haven't noticed before.
* In the log window, when the window is expanded, the text field/view is not expanded with it (unlike with the addresses window, where the text field does expand when the window is).  I have noticed this before but this may be a good opportunity to fix it.
* I'm not certain the shortcut for preferences is necessary; because the app is a menu bar app, it is never actually in the foreground, and thus there is no point I can actually use the command-comma key combination to launch the preferences.  This caused more confusion for me just having it there.

These are all minor details, overall the app functioned as expected still.  

I think you may need to consider the initial user experience though: It may not be immediately evident to a new user what to do.  If you show the preferences window on first launch, the user might not realize that they can close it without quitting the app, and they might not see it in the menu bar.  If you don't show the preferences panel on first launch, the user may not even realize the app is running and might get frustrated until they realize how the app works.  I don't know the best way to approach this from a UX perspective, nor do I have any evidence that this is an actual dilemma, just my thoughts.

**Comment by sneat** [comment]
Could use wording such as "Save and hide" to indicate that the window is hiding, not that the app is closing.

**Comment by SteffeyDev** [comment]
@sneat I feel like that is appropriate, but also a bit misleading, because the button would neither save anything (the fields are saved on edit completion) nor hide (there is a difference in macOS between hiding a window, which you can do by pressing Cmd-H, and closing a window, which is Cmd-W).

That being said I don’t have a better solution, I’m go to re-install some other menu bar apps to see how they handle initial UX.  I think they use a welcome window that says “The app lives up there ^” but I’m not sure about that

**Comment by danielbuechele** [comment]
A simple fix would be to open the settings window on launch.

On Nov 3, 2017, at 4:32 PM, SteffeyDev <notifications@github.com> wrote:

@sneat <https://github.com/sneat> I feel like that is appropriate, but also
a bit misleading, because the button would neither save anything (the
fields are saved on edit completion) nor hide (there is a difference in
macOS between hiding a window, which you can do by pressing Cmd-H, and
closing a window, which is Cmd-W).

That being said I don’t have a better solution, I’m go to re-install some
other menu bar apps to see how they handle initial UX. I think they use a
welcome window that says “The app lives up there ^” but I’m not sure about
that

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/danielbuechele/atemOSC/pull/53#issuecomment-341757137>,
or mute the thread

**Comment by SteffeyDev** [comment]
@danielbuechele Did you read my original comment which describes the potential issues with that approach?

**Comment by danielbuechele** [comment]
Grr, sorry. please disregard my comment ;-)
I won't be able to work on this in the next days, if any of you has time, feel free to continue on this. Otherwise, I'll pick it up later next week.

**Comment by sneat** [comment]
I don't mind the idea of having the initial screen show a bit of info about where the app runs from. Even if it was an extra window that showed and had a "Don't show again" type thing.

**Comment by sneat** [comment]
One downside to this actually is that you can only run a single instance now. I've been involved in events where we've run multiple instances and (slightly painfully) connected and operated multiple ATEMs.

**Comment by SteffeyDev** [comment]
Closing, decided to go a different direction with the UI

**Pull request closed without merging**